### PR TITLE
Fixup e59511c9168f04cb21e85642aa8875c1cb4b4

### PR DESCRIPTION
**Pull request by kroening**
read_goto_binary does not set the config; only read_object_and_link does that.

Fixes issue #2949

### CODE REVIEW

**Review by peterschrammel — APPROVED** [review]
If possible, add the test that exposed the issue.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: e7cc22e584d20ca109b9b8ad363a454f7e42154e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84904634

**Comment by kroening** [comment]
The test is not trivial with our current scripts; in particular, it would fail on anything other than x86/i386.

**Comment by peterschrammel** [comment]
Ok, needs a rebase.

File: src/goto-programs/read_goto_binary.h [hunk 1]
```diff
@@ -19,25 +19,33 @@ class goto_modelt;
 class message_handlert;
 class symbol_tablet;
 
+// Read given goto binary
+// Does not update config
```
Comment (tautschnig) [inline src/goto-programs/read_goto_binary.h]: Could the documentation please be placed with the definition and use Doxygen style?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 2]
```diff
@@ -898,11 +898,13 @@ void goto_instrument_parse_optionst::get_goto_program()
 {
   status() << "Reading GOTO program from `" << cmdline.args[0] << "'" << eom;
 
+  config.set(cmdline);
+
```
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp:901]: Is this actually necessary/useful at all? AFAIK the config will be overwritten below?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 3]
```diff
@@ -898,11 +898,13 @@ void goto_instrument_parse_optionst::get_goto_program()
...
   if(read_goto_binary(cmdline.args[0],
     goto_model, get_message_handler()))
     throw 0;
 
-  config.set(cmdline);
+  config.set_from_symbol_table(goto_model.symbol_table);
```
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: We might need a custom wrapper script to test this, but some tests certainly seem necessary. In particular, the command-line should actually be ignored.

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-instrument/goto_instrument_parse_options.cpp:901]: Yes, but not all of it -- only what is called 'architectural configuration'. Say include paths aren't.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a816370278bfc2e2329af620a9967e5bf1d5194d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85188255

**Comment by tautschnig** [comment]
May I suggest that, in place of a proper CI-ed test, the commit message of the second commit describe in detail a workflow that failed before and works now?

**Comment by kroening** [comment]
ok, done

**Comment by kroening** [comment]
There is temptation to allow regression tests that are arbitrary shell scripts; my main fear is that we'll have tests that nobody can debug anymore.

File: src/goto-programs/read_goto_binary.cpp [hunk 4]
```diff
@@ -27,6 +27,8 @@ Module: Read Goto Programs
 #include "elf_reader.h"
 #include "osx_fat_reader.h"
 
+/// Read given goto binary
+/// Does not update config
```
Comment (tautschnig) [inline src/goto-programs/read_goto_binary.cpp]: Would it be possible to be a bit more explicit? To be frank: this is a commit that has one job, and it doesn't actually do that job. The parameters are not documented, the return value is not explained, failure modes are not covered.

**Review by tautschnig — COMMENTED** [review]
Both ends of the spectrum:
1) I really like how the commit message has been extended (with the minor note that mentioning 32 vs. 64 bit architecture seems necessary). I think this is a better route than coming up with arbitrary test scripts, even though it would mean we don't catch such problems in CI for now.
2) The comments added need a big overhaul.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 6aa7d25bea8c30e2b939d5665169ad503fe9cb06).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85395135

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 637afdc01d0e6fed3b730c935524499e7c4e9eeb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85544888

**Review by tautschnig — APPROVED** [review]

**Pull request merged**